package vdcapi;

import "vdcapi.proto";

enum Type {
    GENERIC_RESPONSE = 1;

    VDSM_REQUEST_HELLO = 2;
    VDC_RESPONSE_HELLO = 3;

    VDSM_REQUEST_GET_PROPERTY = 4;
    VDC_RESPONSE_GET_PROPERTY = 5;

    VDSM_REQUEST_SET_PROPERTY = 6;
    VDC_RESPONSE_SET_PROPERTY = 7;

    VDSM_SEND_PING = 8;
    VDC_SEND_PONG = 9;
    VDC_SEND_ANNOUNCE_DEVICE = 10;
    VDC_SEND_VANISH = 11;
    VDC_SEND_PUSH_PROPERTY = 12;
    VDSM_SEND_REMOVE = 13;
    VDSM_SEND_BYE = 14;
    VDC_SEND_ANNOUNCE_VDC = 23; // new in API v2

    
    VDSM_NOTIFICATION_CALL_SCENE = 15;
    VDSM_NOTIFICATION_SAVE_SCENE = 16;
    VDSM_NOTIFICATION_UNDO_SCENE = 17;
    VDSM_NOTIFICATION_SET_LOCAL_PRIO = 18;
    VDSM_NOTIFICATION_CALL_MIN_SCENE = 19;
    VDSM_NOTIFICATION_IDENTIFY = 20;
    VDSM_NOTIFICATION_SET_CONTROL_VALUE = 21;
    VDSM_NOTIFICATION_DIM_CHANNEL = 24; // new in API v2

    VDC_SEND_IDENTIFY = 22;
}

message Message {
    required Type type = 1 [ default = GENERIC_RESPONSE ];
    optional uint32 message_id = 2 [ default = 0 ];
    optional GenericResponse generic_response = 3;

    // could not be solved via extensions because protobuf-c does
    // not support extensions yet

    // requests that expect specific responses
    optional vdsm_RequestHello vdsm_request_hello = 100;
    optional vdc_ResponseHello vdc_response_hello = 101;

    optional vdsm_RequestGetProperty vdsm_request_get_property = 102;
    optional vdc_ResponseGetProperty vdc_response_get_property = 103;

    // expects generic response
    optional vdsm_RequestSetProperty vdsm_request_set_property = 104;

    // requests that only expect a response in case of error
    optional vdsm_SendPing vdsm_send_ping = 105;
    optional vdc_SendPong vdc_send_pong = 106;
    optional vdc_SendAnnounceDevice vdc_send_announce_device = 107;
    optional vdc_SendVanish vdc_send_vanish = 108;
    optional vdc_SendPushProperty vdc_send_push_property = 109;
    optional vdsm_SendRemove vdsm_send_remove = 110;
    optional vdsm_SendBye vdsm_send_bye = 111;
    optional vdc_SendAnnounceVdc vdc_send_announce_vdc = 120; // new in API v2

    // notifications do not expect any response, not even error responses
    optional vdsm_NotificationCallScene vdsm_send_call_scene = 112;
    optional vdsm_NotificationSaveScene vdsm_send_save_scene = 113;
    optional vdsm_NotificationUndoScene vdsm_send_undo_scene = 114;
    optional vdsm_NotificationSetLocalPrio vdsm_send_set_local_prio = 115;
    optional vdsm_NotificationCallMinScene vdsm_send_call_min_scene = 116;
    optional vdsm_NotificationIdentify vdsm_send_identify = 117;
    optional vdsm_NotificationSetControlValue vdsm_send_set_control_value = 118;
    optional vdsm_NotificationDimChannel vdsm_send_dim_channel = 121; // new in API v2
    optional vdc_SendIdentify vdc_send_identify = 119;
}

enum ResultCode {
    ERR_OK = 0;
    ERR_MESSAGE_UNKNOWN = 1;
    ERR_INCOMPATIBLE_API = 2;
    ERR_SERVICE_NOT_AVAILABLE = 3;
    ERR_INSUFFICIENT_STORAGE = 4;
    ERR_FORBIDDEN = 5;
    ERR_NOT_IMPLEMENTED = 6;
    ERR_NO_CONTENT_FOR_ARRAY = 7;
    ERR_INVALID_VALUE_TYPE = 8;
    ERR_MISSING_SUBMESSAGE = 9;
    ERR_MISSING_DATA = 10;
    ERR_NOT_FOUND = 11;
    ERR_NOT_AUTHORIZED = 12;
}

message GenericResponse {
    required ResultCode code = 1 [ default = ERR_OK ];
    optional string description = 2;
}
